x/payment/keeper/keeper.go

```diff
@@ -32,7 +41,8 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 func (k Keeper) PayForData(goCtx context.Context, msg *types.MsgPayForData) (*types.MsgPayForDataResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	ctx.GasMeter().ConsumeGas(msg.MessageSize, payForDataGasDescriptor)
+	gasToConsume := uint64(shares.MsgSharesUsed(int(msg.MessageSize)) * GasPerMsgShare)
```

Comment:
[comment]
we should use generics or something idk, these type conversion gymnastics are everywhere